### PR TITLE
Add Big Queue implementation

### PR DESCRIPTION
**Pull request by ashish-goswami**
Contains following changes:
* IBigQueue interface for Big Queue
* BigQueue struct implementing IBigQueue
* Test cases for BigQueue functions
* Fix/Refactor index test cases to make more generic

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,3 @@
-# bigqueue [![Build Status](https://travis-ci.com/grandecola/bigqueue.svg?branch=master)](https://travis-ci.com/grandecola/bigqueue)
+# bigqueue [![Build Status](https://travis-ci.com/grandecola/bigqueue.svg?branch=master)](https://travis-ci.com/grandecola/bigqueue) [![Go Report Card](https://goreportcard.com/badge/github.com/grandecola/bigqueue)](https://goreportcard.com/report/github.com/grandecola/bigqueue) [![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT) [![GoDoc](https://godoc.org/github.com/grandecola/bigqueue?status.svg)](https://godoc.org/github.com/grandecola/bigqueue)
 
-BigQueue implementation in Golang
+BigQueue implementation in Golang´
```
Comment (mangalaman93) [inline README.md]: I think there is a extra unnecessary ` at the end of this line

File: bigqueue.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	"fmt"
+	"path"
+)
+
+const (
+	cDataFileFmt  = "data_%d.dat"
+	cDataFileSize = 128 * 1024 * 1024
+)
+
+// BigQueue implements bigqueue interface using Go slices
```
Comment (mangalaman93) [inline bigqueue.go]: The description doesn't look right

File: bigqueue.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+type BigQueue struct {
+	index     *QueueIndex
+	arenaList []*Arena
+	dir       string
+}
+
+// NewBigQueue constructs an instance of bigqueue
```
Comment (mangalaman93) [inline bigqueue.go]: *`BigQueue`

File: bigqueue.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	arenaList := make([]*Arena, 0)
+
+	bq := &BigQueue{
+		index:     index,
+		arenaList: arenaList,
+		dir:       dir,
+	}
+
+	for i := headAid; i <= tailAid; i++ {
+		if err := bq.getNewArena(i); err != nil {
```
Comment (mangalaman93) [inline bigqueue.go]: Given the name of the function uses `get`, it shouldn't update the `arenaList`

Comment (mangalaman93) [reply bigqueue.go]: I think there is no need to call a separate function here!

File: bigqueue.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+	return headAid == tailAid && headOffset == tailOffset
+}
+
+// Peek returns the head of the queue
+func (bq *BigQueue) Peek() ([]byte, error) {
+	aid, offset := bq.index.GetHead()
+
+	// read length
+	aid, offset, lenBytes, err := bq.readBytes(aid, offset, 8)
```
Comment (mangalaman93) [inline bigqueue.go]: We should utilize `ReadUint64` functions of mmap here

File: bigqueue.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+// Enqueue adds a new element to the tail of the queue
+func (bq *BigQueue) Enqueue(message []byte) error {
+	aid, offset := bq.index.GetTail()
+
+	// write length
+	var err error
+	length := uint64(len(message))
+	aid, offset, err = bq.writeBytes(aid, offset, Uint64ToBytes(length))
```
Comment (mangalaman93) [inline bigqueue.go]: We should utilize `WriteUint64` functions of mmap library. This will avoid double copy too.

File: bigqueue.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+			return 0, 0, nil, err
+		}
+		counter += bytesRead
+
+		// check if read all bytes
+		if counter < length {
+			aid++
+			offset = 0
+		} else {
+			// if read bytes were from initial arena and did not include last offset
```
Comment (mangalaman93) [inline bigqueue.go]: we should take care of this in each iteration instead of in the last one

File: bigqueue.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+	for {
+		bytesWritten, err := bq.arenaList[aid].Write(byteSlice[counter:], offset)
+		if err != nil {
+			return 0, 0, err
+		}
+		counter += bytesWritten
+
+		// check if all bytes are written
+		if counter < length {
```
Comment (mangalaman93) [inline bigqueue.go]: we should abstract this part out

File: bigqueue.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+		}
+	}
+
+	return aid, offset, nil
+}
+
+// Flush will unmap all arenas
+func (bq *BigQueue) Flush() {
+	for _, arena := range bq.arenaList {
+		arena.Unmap()
```
Comment (mangalaman93) [inline bigqueue.go]: Why is flush unmapping pages?

File: bigqueue_test.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	if err != nil {
+		t.Errorf("deque failed with err: %s", err)
+	}
+
+	if len(dequeuedMsg) != len(msg) {
+		t.Errorf("lengths does not match, actual: %d, expected: %d", len(dequeuedMsg), len(msg))
+	}
+}
+
+func TestEnqueOverlapLength(t *testing.T) {
```
Comment (mangalaman93) [inline bigqueue_test.go]: Enqueue!

**Comment by zeoaman** [comment]
@ashish-goswami seems like `Flush` function should in fact be called `Close`. And we need to fix the implementation for `Flush`. Also, it would make sense to have `Close` function to the IBigQueue interface.

**Comment by zeoaman** [comment]
I was also thinking if it would be better to store the length as string instead of byte array. Benefits -
* the data file are now human readable
* no need to worry about big endian and little endian

problems -
* will have to use fixed size length for the string array so that we understand the boundaries
* Integer to string would be slower than integer to byte array

**Pull request closed without merging**